packages/browser/src/profiling/lifecycleMode/traceLifecycleProfiler.ts

```diff
@@ -0,0 +1,312 @@
+ * - While running, periodically stops and restarts the JS self-profiling API to collect chunks
+ *
+ * Profiles are emitted as standalone `profile_chunk` envelopes either when:
+ * - there are no more sampled root spans, or
+ * - the 60s chunk timer elapses while profiling is running.
+ */
+export class BrowserTraceLifecycleProfiler {
+  private _client: Client | undefined;
+  private _profiler: JSSelfProfiler | undefined;
+  private _chunkTimer: ReturnType<typeof setTimeout> | undefined;
+  private _activeRootSpanCount: number;
+  // For keeping track of active root spans
+  private _activeRootSpanIds: Set<string>;
+  private _profilerId: string | undefined;
+  private _isRunning: boolean;
+  private _sessionSampled: boolean;
+
+  public constructor() {
+    this._client = undefined;
+    this._profiler = undefined;
+    this._chunkTimer = undefined;
+    this._activeRootSpanCount = 0;
+    this._activeRootSpanIds = new Set<string>();
+    this._profilerId = undefined;
+    this._isRunning = false;
+    this._sessionSampled = false;
+  }
+
+  /**
+   * Initialize the profiler with client and session sampling decision computed by the integration.
+   */
+  public initialize(client: Client, sessionSampled: boolean): void {
+    DEBUG_BUILD && debug.log("[Profiling] Initializing profiler (lifecycle='trace').");
+
+    this._client = client;
+    this._sessionSampled = sessionSampled;
+
+    client.on('spanStart', span => {
+      if (span !== getRootSpan(span)) {
+        return;
+      }
+      if (!this._sessionSampled) {
+        DEBUG_BUILD && debug.log('[Profiling] Session not sampled because of negative sampling decision.');
+        return;
+      }
+      // Only count sampled root spans
+      if (!span.isRecording()) {
+        DEBUG_BUILD && debug.log('[Profiling] Discarding profile because root span was not sampled.');
+        return;
+      }
+
+      const rootSpanJSON = spanToJSON(span);
+      const spanId = rootSpanJSON.span_id as string | undefined;
+      if (!spanId) {
+        return;
+      }
+      if (this._activeRootSpanIds.has(spanId)) {
+        return;
+      }
+      this._activeRootSpanIds.add(spanId);
+
+      const wasZero = this._activeRootSpanCount === 0;
+      this._activeRootSpanCount++; // Increment before eventually starting the profiler
+      DEBUG_BUILD &&
+        debug.log(
+          `[Profiling] Root span ${rootSpanJSON.description} started. Active root spans:`,
+          this._activeRootSpanCount,
+        );
+      if (wasZero) {
+        this.start();
+      }
+    });
+
+    client.on('spanEnd', span => {
+      if (!this._sessionSampled) {
+        return;
+      }
+
+      const spanJSON = spanToJSON(span);
+      const spanId = spanJSON.span_id as string | undefined;
+      if (!spanId || !this._activeRootSpanIds.has(spanId)) {
+        return;
+      }
+
+      this._activeRootSpanIds.delete(spanId);
+      this._activeRootSpanCount = Math.max(0, this._activeRootSpanCount - 1);
```

Comment:
I would add a debugLog here where you check that
```
if debug && activeRootSpanCount - 1 < 0 
	we have an SDK bug, span tracking is broken
```

Reply:
I deleted `_activeRootSpanCount` now to just use `this._activeRootSpanIds.size`. I figured, I don't need the extra tracking.